### PR TITLE
Use VCR.py for testing

### PR DESCRIPTION
**Pull request by joshtemple**
This PR represents a large refactor of the test suite for Spectacles. Namely, it solves a big problem with our current test suite: we depend a lot on heavy mocking for external requests. Since Spectacles fundamentally works off the Looker API, its functionality is very dependent upon the responses from various API endpoints.

Previously, we mocked fake data for most of these calls, which is brittle because API response formats can change and mistakes can be made when mocking.

VCR.py represents a different paradigm: it allows you to record HTTP requests and responses to serialized YAML files called cassettes which are stored in version control. With certain configurations, VCR.py can prevent tests from making new requests, instead "replaying" the results of the cassettes. This is a much better solution because if the Looker API changes, we can simply record new cassettes and update our code. Similarly, we can see the full responses and requests made by our code.

There are some caveats and instructions, which I've updated in CONTRIBUTING.md. Additionally, this adds a greater learning curve for contribution, so we'll have to support contributors who want to write tests that leverage this new library.

### CODE REVIEW

File: tests/test_sql_validator.py [hunk 1]
```diff
@@ -1,217 +1,191 @@
...
+from spectacles.exceptions import SpectaclesException
+from spectacles.lookml import Dimension
+
+EXPECTED_QUERY_COUNTS = {"models": 1, "explores": 1, "dimensions": 5}
+
+
+@pytest.fixture(scope="class")
+def validator(looker_client, record_mode) -> Iterable[SqlValidator]:
+    with vcr.use_cassette(
+        f"tests/cassettes/test_sql_validator/fixture_validator_init.yaml",
```
Comment (DylanBaker) [inline tests/test_sql_validator.py:16]: Errant f-string I think.

**Comment by DylanBaker** [comment]
I realise that my comment above was actually stupid. Because we hard code our URL, contributors can't run their own.

File: tests/test_client.py [hunk 2]
```diff
@@ -53,190 +56,165 @@ def client_kwargs():
...
 
-@pytest.fixture
-def client(monkeypatch):
-    mock_authenticate = Mock(spec=LookerClient.authenticate)
-    monkeypatch.setattr(LookerClient, "authenticate", mock_authenticate)
-    return LookerClient(TEST_BASE_URL, TEST_CLIENT_ID, TEST_CLIENT_SECRET)
+@pytest.mark.vcr(match_on=["uri", "method", "raw_body"])
+def test_branch_management_should_work(looker_client):
+    project = "eye_exam"
+    tmp_branch = f"tmp-pytest"
```
[the remark this answers is not in the document]
Comment (joshtemple) [reply tests/test_client.py:62]: Absolutely not lol

**Comment by joshtemple** [comment]
I think the solution is to publish the `eye_exam` repo so people can see what we're testing against. I don't really see a strong reason for leaving it as a private repo

File: tests/test_client.py [hunk 3]
```diff
@@ -53,190 +56,165 @@ def client_kwargs():
...
+#     )
+
+
+# @patch("spectacles.client.requests.Session.delete")
+# def test_delete_branch(mock_delete, client):
+#     client.delete_branch("project_name", "test_branch_name")
+#     mock_delete.assert_called_once_with(
+#         url="https://test.looker.com:19999/api/3.1/projects/project_name/git_branch/test_branch_name",
+#         timeout=300,
+#     )
```
[the remark this answers is not in the document]
Comment (joshtemple) [reply tests/test_client.py:220]: Whoops! I'll take care of this.

**Comment by DylanBaker** [comment]
@joshtemple That makes sense. I imagine we'll need a few branches for a variety of setups? i.e. one with a manifest.lkml file.

**Pull request merged**